src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -28,6 +29,7 @@ public abstract partial class Renderer : IDisposable, IAsyncDisposable
     {
         private readonly IServiceProvider _serviceProvider;
         private readonly Dictionary<int, ComponentState> _componentStateById = new Dictionary<int, ComponentState>();
+        private readonly Dictionary<IComponent, ComponentState> _componentStateByComponent = new Dictionary<IComponent, ComponentState>();
```

Comment:
I understand this dictionary is to facilitate the look up in the DispatchEventAsync and so on. 
Would iterating over the ComponentState entries instead work?

The reason I'm bringing this up is because it implies more allocation and housekeeping we need to perform whether an exception happens or not, and by the time an exception happens (by definition) we are already in an "exceptional" state, so we shouldn't care about perf in that case.

Reply:
From inspection this is only used in the event handler, so I think it might be better to avoid this dictionary all together and just iterate over the list of component state entries

Reply:
Unfortunately that's not the case. We have to find the component state before beginning event dispatch in case it got disposed before the event handler completed, so we need to do this lookup on every incoming event.

Reply:
Ahh, I see. Is it because of the async case where the error boundary might get disposed before the exception happens?

Reply:
Yes, or if the receiver component gets disposed before its event handler raises an async exception, which is potentially pretty common (e.g., it happens every time if you have multiple children and more than one have async operations that fail, because the first error will cause the disposal of both, then the second error gets received).